### PR TITLE
Limit the allowed epoch timestamps

### LINKED ISSUE #3444: Bare integers are accepted for due: and wait:, but cause issues later
#### To report a bug...

* What command(s) did you run?

`task add recur:daily due:10 wait:6 foobar`

* What did you expect to happen?

The goal was to add a daily task that recurs at 10am.

* What actually happened?

`task list` hangs and thousands of tasks are created while it hangs.

Screenshot 2024-05-06 at 3 27 22 PM

* Paste the output of the ```task diag``` command.

```
task 3.0.2
   Platform: Darwin

Compiler
    Version: Apple LLVM 15.0.0 (clang-1500.3.9.4)
       Caps: +stdc +stdc_hosted +LP64 +c8 +i32 +l64 +vp64 +time_t64
 Compliance: C++17

Build Features
      CMake: 3.29.2
    libuuid: libuuid + uuid_unparse_lower
 Build type: Release

Configuration
       File: /Users/jerry/.taskrc (found), 1469 bytes, mode 100644
       Data: /Users/jerry/.task (found), dir, mode 40755
    Locking: Enabled
         GC: Enabled
    $EDITOR: nvim
Hooks
     System: Enabled
   Location: /Users/jerry/.task/hooks
             (-none-)

Tests
   Terminal: 252x117
       Dups: Scanned 8450 tasks for duplicate UUIDs:
             No duplicates found
 Broken ref: Scanned 8450 tasks for broken references:
             No broken references found
```
---

Running `task delete` will go through the tasks one by one and delete them but it takes forever if there are thousands of them. What would the effect be of deleting the rows in the `tasks` table? I didn't try this because I see there are other tables referencing that data, but could that be a 'last resort' way of recovering?

Thanks!

### PR DESCRIPTION
**Pull request by lanker**
The code for parsing epoch timestamps when displaying tasks only supports values between year 1980 and 9999. Previous to this change, it was possible to set e.g., the due timestamp to a value outside of these limits, which would make it impossible to later on show the task.

With this change, we only allow setting values within the same limits used by the code for displaying tasks.

Fixes https://github.com/GothenburgBitFactory/taskwarrior/issues/3444

### CODE REVIEW

**Comment by lanker** [comment]
This PR is depending on https://github.com/GothenburgBitFactory/libshared/pull/83. Not sure how you handle bumping the version of libshared that is used in Taskwarrior...?

**Comment by djmitche** [comment]
Let's wait until that PR is merged, and then include a change to the submodule version in this PR.

That will also affect https://github.com/GothenburgBitFactory/taskwarrior/issues/3623 but I think there's no problem in rolling both of those at once.

**Comment by lanker** [comment]
Just some additional info if anyone ends up here in the future:

Looking at the commit and PR history, I couldn't find any reason for the lower limit when [parsing timestamps](https://github.com/GothenburgBitFactory/libshared/blob/a63fa350413f598b3b55355a6460b7fd2efb2e69/src/Datetime.cpp#L697) (the upper limit was introduced from [this issue](https://github.com/GothenburgBitFactory/taskwarrior/issues/2373)), so my initial approach was to remove it completely. However, it seems like more strings than potential epoch timestamps are sent to `Datetime::parse_epoch`.

For example when doing `task 3 info`, `Datetime::parse_epoch` is called with the value `3`, and must return `false` for it to work. So somehow, we need to differentiate between numbers being a timestamp and numbers being something else, like a task ID. I think that's why there's a lower limit. I guess it would've been nice to only trying to parse strings that we know should be timestamps, but I'm sure there's a good reason for it to work as it does.

(So there might be problems if a task ID would be > 315532800, but let's hope no one ends up there... :))

**Comment by djmitche** [comment]
OK, the libshared update has been merged into Taskwarrior, if you'd like to rebase this PR.

**Comment by djmitche** [comment]
If you're ready for a review on this, please add me and/or @smemsh as a reviewer!

**Comment by lanker** [comment]
> If you're ready for a review on this, please add me and/or @smemsh as a reviewer!

I don't think I have enough permissions to be able to add reviewers. According to the [documentation](https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/proposing-changes-to-your-work-with-pull-requests/requesting-a-pull-request-review): "To assign a reviewer to a pull request, you will need write access to the repository.".

So please add yourself and anyone else that should have a look at it, thanks!

**Review by smemsh — COMMENTED** [review]
Have you considered just using `Datetime::parse_epoch()` itself and checking the return value? Since the range check is already in there.  This way it wouldn't have to be duplicated in both taskwarrior and libshared.  Just an idea...

**Comment by smemsh** [comment]
> Have you considered just using `Datetime::parse_epoch()` itself

But since you're using those defines, and it's just a simple bounds check, it seems valid enough in any case to do it like your current patch already does.

So, the code looks sane to me @djmitche ...

**Comment by lanker** [comment]
> Have you considered just using `Datetime::parse_epoch()` itself and checking the return value? Since the range check is already in there. This way it wouldn't have to be duplicated in both taskwarrior and libshared. Just an idea...

That would actually be a better solution, so I took a brief look at it, not much luck though... Firstly, `Datetime::parse_epoch()` is private to `Datetime`, and I'm not comfortable enough with the code (or C++ in general) to start changing visibilities. Secondly, `Datetime::parse_epoch()` takes a `Pig` as argument. To create a `Pig` you need a `string`, I couldn't find any string that would work for us here.

I think it could've been nice to have a `bool Datetime::check_epoch(time_t epoch)` that could be used both here and in libshared, but might be a bit overkill for just a bounds check.

**Review by djmitche — APPROVED** [review]

**Pull request merged**
